rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc

```diff
@@ -0,0 +1,275 @@
+import analysis::typepal::TypePal;
+import analysis::typepal::TModel;
+
+import analysis::diff::edits::TextEdits;
+
+import vis::Text;
+
+import util::Maybe;
+import util::Reflective;
+
+// Only needed until we move to newer type checker
+// TODO Remove
+import ValueIO;
+
+data IllegalRenameReason
+    = InvalidName(str name)
+    | CausesDoubleDeclaration(rel[loc, loc] defs)
+    | CausesCapture(rel[loc use, loc oldDef, loc newDef] captures)
+    | DeclarationsNotEditable(set[loc] decls)
+    ;
+
+data RuntimeException
+    = IllegalRename(loc location, IllegalRenameReason reason)
+    | UnsupportedRename(set[tuple[loc location, str message]])
+    | UnexpectedFailure(str message)
+;
+
+private set[loc] getUses(TModel tm, loc def) {
+    return invert(getUseDef(tm))[def];
+}
+
+private set[Define] getDefines(TModel tm, loc useDef) {
+    set[loc] defLocs = tm.useDef[useDef] != {}
+                     ? tm.useDef[useDef] // `useDef` is a use location
+                     : {useDef};         // `useDef` is a definition location
+
+    return {tm.definitions[d] | d <- defLocs};
+}
+
+void throwIfNotEmpty(&T(&A) R, &A arg) {
+    if (arg != {}) {
+        throw R(arg);
+    }
+}
+
+void checkLegalName(str name) {
+    try {
+        parse(#Name, escapeName(name));
+    } catch ParseError(_): {
+        throw InvalidName(name);
+    }
+}
+
+void checkCausesDoubleDeclaration(set[Define] currentDefs, set[Define] newDefs) {
+    // Is newName already resolvable from a scope where <current-name> is currently declared?
+    throwIfNotEmpty(CausesDoubleDeclaration, {<nD.defined, cD> | Define nD <- newDefs
+                                                               , loc cD <- currentDefs.defined
+                                                               , isContainedIn(cD, nD.scope)
+                                             });
+}
+
+bool isImplicitDefinition(start[Module] _, rel[loc use, loc def] useDef, Define _: <_, _, _, variableId(), defined, _>) = defined in useDef.use; // use of name at implicit declaration loc
+
+bool isImplicitDefinition(start[Module] m, rel[loc, loc] _, Define _: <_, _, _, patternVariableId(), defined, _>) {
+    visit (m) {
+        case qualifiedName(qn):
+            if (locationOfName(qn) == just(defined)) return true;
+        case multiVariable(qn):
+            if (locationOfName(qn) == just(defined)) return true;
+        case variableBecomes(n, _):
+            if (locationOfName(n) == just(defined)) return true;
+    }
+
+    return false;
+}
+
+default bool isImplicitDefinition(start[Module] _, rel[loc, loc] _, Define _) = false;
+
+void checkCausesCapture(TModel tm, start[Module] m, set[Define] currentDefs, set[loc] currentUses, set[Define] newDefs) {
+    set[Define] newNameImplicitDefs = {def | Define def <- newDefs, isImplicitDefinition(m, tm.useDef, def)};
+    set[Define] newNameExplicitDefs = {def | Define def <- newDefs, !isImplicitDefinition(m, tm.useDef, def)};
```

Comment:
nit: these two seem like set difference:

```
allDefs = {def | Define def <- newDefs };
newNameImplicitDefs = {d | d <- allDefs, isImplicitDefinition(m, tm.useDef, d)};
newNameExplicitDefs = allDefs - newNameImplicitDefs ;
```

and ideally (also faster):

```
newNameImplicitDefs = findImplicitDefs(m, tm.useDef); // only a single visit over the tree instead of a visit over the tree for every define
newNameExplicitDefs = {def | Define def <- newDefs } - newNameImplicitDefs;
```

Reply:
This is not purely a visit over the tree. In case of pattern variables, this syntactical analysis is needed, but for regular variables, it is not. However, I understand the point and can rewrite to something that visits the tree at most once.